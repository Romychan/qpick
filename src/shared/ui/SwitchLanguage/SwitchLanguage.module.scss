@import '~/app/assets/styles/_typography.scss';

.container {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.icon {
  stroke: var(--gray-600) !important;
}

.button {
  &.active {
    @include text-bold;
    color: var(--yellow-500);
  }
}
