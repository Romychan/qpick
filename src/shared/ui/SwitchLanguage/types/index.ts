/**
 * Base language item interface
 *
 * @template T The type of the unique value
 */
export interface LanguageOption<T = string> {
  /** The unique value of the element */
  value: T;
  /** The displayed value of the element */
  label: string;
}
