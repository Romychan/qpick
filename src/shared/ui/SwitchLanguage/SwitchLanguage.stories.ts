import type { Meta, StoryObj } from '@storybook/react';

import { SwitchLanguage } from './SwitchLanguage';

const meta = {
  title: 'UI/SwitchLanguage',
  component: SwitchLanguage,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof SwitchLanguage>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    languages: [
      {
        value: 'ru',
        label: 'RU',
      },
      {
        value: 'en',
        label: 'EN',
      },
    ],
  },
};
