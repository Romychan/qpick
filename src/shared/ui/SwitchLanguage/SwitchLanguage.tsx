import { useState } from 'react';

import { cl } from '../../lib/utils';
import { Button } from '../Button';
import { Icon } from '../Icon';

import { LanguageOption } from './types';
import styles from './SwitchLanguage.module.scss';

interface ISwitchLanguageProps {
  /** List of languages */
  languages: LanguageOption[];
}

/** This component is used to switch the language on the page */
export const SwitchLanguage = ({ languages }: ISwitchLanguageProps) => {
  const [currentLanguage, setCurrentLanguage] = useState(languages[0].value);

  return (
    <div className={styles.container}>
      <Icon name="planet" className={styles.icon} size={20} />
      {languages.map(({ value, label }) => (
        <Button
          key={value}
          size="sm"
          variant="link-primary"
          className={cl(styles.button, {
            [styles.active]: currentLanguage === value,
          })}
          onClick={() => setCurrentLanguage(value)}
          data-testid="language-button"
        >
          {label}
        </Button>
      ))}
    </div>
  );
};
