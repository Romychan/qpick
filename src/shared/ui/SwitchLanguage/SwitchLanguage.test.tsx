import {
  RenderResult,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import { SwitchLanguage } from './SwitchLanguage';

const MOCK_LANGUAGES = [
  {
    value: 'ru',
    label: 'RU',
  },
  {
    value: 'en',
    label: 'EN',
  },
];

describe('SwitchLanguage', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = render(<SwitchLanguage languages={MOCK_LANGUAGES} />);
  });

  it('should render correctly', () => {
    const buttonElement = screen.getAllByTestId('language-button');

    expect(buttonElement).toHaveLength(2);
  });

  it('should switch languages correctly', () => {
    const buttonFirstElement = screen.getByRole('button', { name: 'RU' });
    const buttonSecondElement = screen.getByRole('button', { name: 'EN' });

    expect(buttonFirstElement.getAttribute('class')).toMatch('active');
    expect(buttonSecondElement.getAttribute('class')).not.toMatch('active');

    fireEvent.click(buttonSecondElement);

    expect(buttonFirstElement.getAttribute('class')).not.toMatch('active');
    expect(buttonSecondElement.getAttribute('class')).toMatch('active');
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
