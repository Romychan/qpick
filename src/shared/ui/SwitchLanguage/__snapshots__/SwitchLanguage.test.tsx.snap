// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SwitchLanguage > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_b93fe0"
      >
        <svg
          class="icon icon-planet _icon_b93fe0"
          data-testid="icon"
          height="20"
          width="20"
        >
          <use
            xlink:href="/images/icons.svg#planet"
          />
        </svg>
        <button
          class="_button_0302f3 _button_b93fe0 _active_b93fe0 _link-primary_0302f3 _sm_0302f3"
          data-testid="language-button"
          type="button"
        >
          RU
        </button>
        <button
          class="_button_0302f3 _button_b93fe0 _link-primary_0302f3 _sm_0302f3"
          data-testid="language-button"
          type="button"
        >
          EN
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_b93fe0"
    >
      <svg
        class="icon icon-planet _icon_b93fe0"
        data-testid="icon"
        height="20"
        width="20"
      >
        <use
          xlink:href="/images/icons.svg#planet"
        />
      </svg>
      <button
        class="_button_0302f3 _button_b93fe0 _active_b93fe0 _link-primary_0302f3 _sm_0302f3"
        data-testid="language-button"
        type="button"
      >
        RU
      </button>
      <button
        class="_button_0302f3 _button_b93fe0 _link-primary_0302f3 _sm_0302f3"
        data-testid="language-button"
        type="button"
      >
        EN
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
